### ISSUE TITLE
RFC: most charts should be StatefullSet, but are Deployments

### ISSUE THREAD

**Issue by ishioni** [billimek/billimek-charts#295]
I believe that charts which deploy stateful applications, like for example Home Assistant, radarr et all should deploy statefullsets instead of deployments. I know that one can get almost the same result using a chart hardcoded replica number, and deployment strategy so as to guard from 2 copies accessing the same file (like on a new version rollout) from a PVC. I also realise that most of these applications have no business of being deployed in more than one copy from the same deployment, since they don't cluster in any meaningful way.

Meanwhile statefullsets have these things built-in. They also have a neat optional bonus (or a hindrance, depending on ones usecase), that because they bring pvc templates instead of pvc, helm will not delete the pvc outright on uninstall.

This is not a critique, it's a call for comments: are people using deployments instead of statefullsets because of familiarity, or is there something i'm not seeing?

**Comment by billimek** [comment]
I'm for it, @ishioni.  Back when some of these charts were hosted in the [helm/charts](https://github.com/helm/charts) repo, some of us raised this question and there was a lot of debate.  It's worth referencing that conversation here as well:

* https://github.com/helm/charts/issues/1863
* https://github.com/helm/charts/pull/12047 (read the comments in this one)

**Comment by billimek** [comment]
If we choose to pursue this, obviously it would only be necessary to alter charts that basically have a requirement to use persistent storage in order to function properly (e.g. radarr/sonarr/plex/etc).  A change from deployment to statefulset will also be a 'breaking' change and will require a major version change as well as documented upgrade path.

**Comment by ishioni** [comment]
After reading both of those threads i still think we should proceed forward with going to STS when PVCs are involved. I totally agree with @unguiculus that STS is not a magic fix for the types of applications which don't scale at all, however @davidkarlsen rightly points out that it has some perks over normal DPs.

I think the best thing to do is to point out somewhere in new PRs that stateful applications should now uses STS with claimTemplates instead of DPs with PVCs. This repo has relatively less charts than stable, so I thin it'd be a good idea not to exacerbate the issue. 

>as well as documented upgrade path

This has me stumped most of all. When i converted [my charts](https://gitlab.com/inzmovi/helm-charts), most of the time it was best to either migrate the data out-of-bound or use a transporter pod with both PVCs mounted. I *think* the best way would be to point users to use existingClaim? But then we're still up against helm managing the lifecycle of the PVC, and that may very well lead to deletion

**Comment by onedr0p** [comment]
It might be worth using our k8s-at-home group to host charts, maybe we can carry ones over there as STS and slowly deprecate this repo?

https://github.com/k8s-at-home

**Comment by tabnul** [comment]
Why would i want a statefullset, if i use kubernetes native replicated storage such as longhorn?
Statefulsets are in my opinion just more of a hassle in regards of high availability (not failing over automatically)

**Comment by onedr0p** [comment]
Using Longhorn complicates restoring backups when using stateful sets?

**Comment by tabnul** [comment]
> Longhorn complicates restoring backups when using stateful sets?

I dont know but i just dont see the purpose of running  hass in a statefullset.  Taking into account that you cripple auto failover functionality which is imo a key benefit of running hass in kubernetes to begin with.

**Comment by onedr0p** [comment]
I'm not sure I follow, statefulsets can failover too.

**Comment by tabnul** [comment]
Yes it can failover but needs manual intervention on node failure.
https://github.com/kubernetes/kubernetes/issues/74689

Well, longhorn is not directly related. What it does offer is super flexible distributed storage with easy automatic failovering over nodes. That contradicts for me with the HA characteristics of a statefulset

**Comment by onedr0p** [comment]
According to the following issue this can be solved with configuration options we can define in the chart.

https://github.com/kubernetes/kubernetes/issues/74947#issuecomment-509584054

**Comment by ryholt** [comment]
Looks like it'd be a kube-api arg: https://github.com/kubernetes/kubernetes/issues/74651

**Comment by carpenike** [comment]
Looks like it'd be a kube api arg: https://github.com/kubernetes/kubernetes/issues/74651

**Comment by DirtyCajunRice** [comment]
Just to udpate you  @ishioni i have implemented a configurable deployment/statefulset choice and it is already in the bitwardenrs chart. This can be applied to any/all charts that statefulsets apply to in scope.
